src/relayclient/GSNConfigurator.ts

```diff
@@ -47,19 +48,48 @@ export function configureGSN (partialConfig: Partial<GSNConfig>): GSNConfig {
  * @param partialConfig
  */
 export async function resolveConfigurationGSN (provider: provider, partialConfig: Partial<GSNConfig>): Promise<GSNConfig> {
+  // @ts-ignore
+  if (provider.send == null && provider.sendAsync == null) {
+    throw new Error('First param is not a web3 provider')
+  }
+
   if (partialConfig.relayHubAddress != null) {
     throw new Error('Resolve cannot override passed values')
   }
   if (partialConfig.paymasterAddress == null) {
     throw new Error('Cannot resolve GSN deployment without paymaster address')
   }
+
   const contractInteractor = new ContractInteractor(provider, defaultGsnConfig)
   const paymasterInstance = await contractInteractor._createPaymaster(partialConfig.paymasterAddress)
-  const relayHubAddress = await paymasterInstance.getHubAddr()
+
+  const [
+    chainId, relayHubAddress, forwarderAddress
+  ] = await Promise.all([
+    partialConfig.chainId ?? await contractInteractor.getAsyncChainId(),
+    await paymasterInstance.getHubAddr().catch(e => { throw new Error('Not a paymaster contract') }),
+    partialConfig.forwarderAddress ?? await paymasterInstance.trustedForwarder().catch(e => { throw new Error('paymaster has no trustedForwarder()') }),
+    await paymasterInstance.versionPaymaster().catch((e: any) => { throw new Error('Not a paymaster contract') }).then((version: string) => contractInteractor._validateVersion(version))
```

Comment:
Don't you await on the promises here at the moment of array creation (one-by-one) and then pass resolved promises to the `Promise.all`?

Also, please please create individual promise variables and then await on them. This code is very hard to read.